src/main/scala/mesosphere/marathon/core/task/termination/impl/KillStreamWatcher.scala

```diff
@@ -36,8 +43,9 @@ object KillStreamWatcher extends StrictLogging {
         classOf[UnknownInstanceTerminated], eventStream, bufferSize, overflowStrategy).map(_.id)
 
     // eagerComplete allows us to toss the right materialized cancellable value
-    killedViaInstanceChanged.
-      merge(killedViaUnknownInstanceTerminated, eagerComplete = true)
+    val liveInstanceIds = killedViaInstanceChanged.merge(killedViaUnknownInstanceTerminated, eagerComplete = true)
+
+    Source(alreadyConsideredTerminal).concatMat(liveInstanceIds)(Keep.right)
```

Comment:
What is this doing? Why not simply filter the `liveInstanceIds` with `considerTerminal`?

Reply:
If we're waiting for the unreachable instance, we will have to wait until those instances will be reconciled, so it will be ~10 minutes in the worst case. That optimisation reduces the waiting time if the instance is already unreachable.

Reply:
I'm sorry. I'm slow today and don't get what this line is doing. How are `alreadyConsideredTerminal` and `liveInstanceIds` related? Isn't `concatMat` just chaining both?

Reply:
Yes. Before subscribing to the live stream we push what we already know. It covers the case where we were watching already unreachable instance.

Reply:
I still don't get it 😄. What kind of instance ids are included in `alreadyConsideredTerminal` and `liveInstanceIds`? If `liveInstanceIds` is a subset I don't understand why we concatenate them.

You are right about the advantage but it seems ti complicated things for the sake of early optimization. Come to think it is odd the we wait for an event. It might be a race condition. Also, if an instance is already terminated why should we wait at all?

Reply:
> What kind of instance ids are included in alreadyConsideredTerminal and liveInstanceIds?

`alreadyConsideredTerminal` and `liveInstanceIds` produce the same instanceIds.

> If liveInstanceIds is a subset I don't understand why we concatenate them.

Consider the following case: If the instance is already in the terminal state at the moment of watching, then we won't see any updates in the `liveInstancesIds` except for the updates from the reconciliation process (which might take up to 10 minutes to arrive). That's why we concatenate the live stream with the id of instances which are already known to be in the terminal state.

> You are right about the advantage but it seems ti complicated things for the sake of early optimization.

If we don't do this, users will have to wait up to 10 minutes before the delete instance call will give a response back, which will result in request timeout and make marathon users unhappy. For me it doesn't look like `for the sake of early optimization.`

> Also, if an instance is already terminated why should we wait at all?

That's the whole point for this change: if the instance is already terminated we don't wait for the reconciliation to happen and immediately push instance ids downstream.

Reply:
You convinced me 🙂. Now I just have to understand why we `merge` and then `concat`. That is odd to me 🤔

Reply:
as well as cancellable is materialized properly - it's fine :)

Reply:
I think
```scala
Source(alreadyConsideredTerminal)
  .merge(killedViaInstanceChanged)
  .merge(killedViaUnknownInstanceTerminated)
```

is easier to follow.